### ISSUE TITLE
`offset` intrinsic doesn't check that pointers stay within bounds

### ISSUE THREAD

**Issue by adpaco-aws** [model-checking/kani#1233]
The [documentation for the `offset` method](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset) mentions in the safety conditions that:

> Both the starting and resulting pointer must be either in bounds or one byte past the end of the same [allocated object](https://doc.rust-lang.org/std/ptr/index.html#allocated-object).

At present, the standard library prevents us from adding this check because otherwise we run into spurious failures (i.e., offsets that go out-of-bounds,  are frequently used).

**Comment by adpaco-aws** [comment]
Removing the soundness label since this is not strictly a soundness issue.

Adding this check would introduce many failures that aren't part of the user code, because offset operations often go out-of-bounds (in the standard library and other projects). Let me illustrate with an example in C:

```c
int *p = malloc(10);
p = p + 50; // pointer has been offset 50 units (out of bounds) - not an error since the pointer is not used
int x = *p // this is an error, the index points to somewhere out of the bounds for the object
```

Rust allows manipulations as in line (2), so there's no soundness concerns about that. On the other hand, de-referencing the pointer in line (3) should raise an error, which Kani does.

**Comment by giltho** [comment]
> i.e., offsets that go out-of-bounds, are frequently used
Do you have an example of that? The stdlib is tested with MIRI I believe, and MIRI detects creating OOB pointers as UB.

> de-referencing the pointer in line (3) should raise an error, which Kani does
Not always, see https://github.com/model-checking/kani/issues/1665

**Comment by celinval** [comment]
I think we need to revisit these cases. This is a potential soundness issue since Kani would not be able to detect a pointer going out of bounds. From what I can recall, the spurious CEX would only happen for cases where the offset started from a dangling pointer and the count was 0, which is what should handle. Note that the [`offset` documentation](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset) has been updated:
> If the computed offset is non-zero, then self must be derived from a pointer to some [allocated object](https://doc.rust-lang.org/std/ptr/index.html#allocated-object) (...)
